wallet-api/src/Ledger/Value/TH.hs

```diff
@@ -87,80 +92,34 @@ similar to 'Ledger.Ada' for their own currencies.
 
 -}
 
-type CurrencyMap v = [(CurrencySymbol, v)]
-
-cmap :: Q (TExp ((v -> w) -> CurrencyMap v -> CurrencyMap w))
-cmap = [||
-    \f ->
-        let go [] = []
-            go ((c, i):xs') = (c, f i) : go xs'
-        in go
-    ||]
-
-lookup :: Q (TExp (CurrencySymbol -> CurrencyMap v -> Maybe v))
-lookup = [||
-    \(CurrencySymbol cur) xs ->
-        let
-            go :: [(CurrencySymbol, v)] -> Maybe v
-            go []                          = Nothing
-            go ((CurrencySymbol c, i):xs') = if $$(P.eq) c cur then Just i else go xs'
-        in go xs
- ||]
-
 -- | Get the quantity of the given currency in the 'Value'.
 valueOf :: Q (TExp (Value -> CurrencySymbol -> Int))
-valueOf = [||
-  \(Value xs) cur ->
-      case $$(lookup) cur xs of
+valueOf = [|| 
+  \(Value mp) cur -> 
+      case $$(Map.lookup) $$(eqCurSymbol) cur mp of
         Nothing -> 0 :: Int
         Just i  -> i
    ||]
 
-data These a b = This a | That b | These a b
-
-these :: Q (TExp (a -> b -> (a -> b -> c) -> These a b -> c))
-these = [||
-    \a' b' f -> \case
-        This a -> f a b'
-        That b -> f a' b
-        These a b -> f a b
-    ||]
-
-union :: Q (TExp (CurrencyMap a -> CurrencyMap b -> CurrencyMap (These a b)))
-union = [||
-    \ls rs ->
-        let
-            f a b' = case b' of
-                Nothing -> This a
-                Just b  -> These a b
-            ls' = $$(P.map) (\(c, i) -> (c, (f i ($$(lookup) c rs)))) ls
-            rs' = $$(P.filter) (\(CurrencySymbol c, _) -> $$(P.not) ($$(P.any) (\(CurrencySymbol c', _) -> $$(P.eq) c' c) ls)) rs
-            rs'' = $$(P.map) (\(c, b) -> (c, (That b))) rs'
-        in $$(P.append) ls' rs''
-  ||]
-
 -- | Make a 'Value' containing only the given quantity of the given currency.
 singleton :: Q (TExp (CurrencySymbol -> Int -> Value))
-singleton = [|| \c i -> Value [(c, i)] ||]
+singleton = [|| \c i -> Value ($$(Map.singleton) c i) ||]
 
 unionWith :: Q (TExp ((Int -> Int -> Int) -> Value -> Value -> Value))
 unionWith = [||
-  \f (Value ls) (Value rs) ->
-    let
-        combined = $$(union) ls rs
+  \f (Value ls) (Value rs) -> 
+    let 
+        combined = $$(Map.union) $$(eqCurSymbol) ls rs
         unThese k = case k of
-            This a    -> f a 0
-            That b    -> f 0 b
-            These a b -> f a b
-    in Value ($$(cmap) unThese combined)
+            Map.This a    -> f a 0
+            Map.That b    -> f 0 b
+            Map.These a b -> f a b
```

Comment:
This reads weirdly, but I see the usefulness of re-exporting `These` from `Map`.